### PR TITLE
Validation improvements

### PR DESCRIPTION
**Pull request by Shane32**
Makes many validation rules allocation-free instances

### CODE REVIEW

**Comment by Shane32** [comment]
I finished my code.  Feel free to add the improvements you wanted to.

**Comment by sungam3r** [comment]
Well I see that the idea of this PR the same as #2134. I see that in some places you don't use statics - where there is state. I went further and transferred the state to `ValidationContext`.

File: src/GraphQL/Validation/TypeInfo.cs [hunk 1]
```diff
@@ -34,7 +33,14 @@ public TypeInfo(ISchema schema)
         /// <returns></returns>
         public INode[] GetAncestors()
         {
-            return _ancestorStack.Skip(1).Reverse().ToArray();
+            var ancestorArray = _ancestorStack.ToArray();
+            var c = ancestorArray.Length;
+            var ret = new INode[c - 1];
+            for (int i = c-- - 1; i > 0; --i) {
```
Comment (sungam3r) [inline src/GraphQL/Validation/TypeInfo.cs:39]: 😮

File: src/GraphQL/Validation/ValidationContext.cs [hunk 2]
```diff
@@ -72,15 +72,12 @@ public List<VariableUsage> GetVariables(IHaveSelectionSet node)
...
-                    varRef => usages.Add(new VariableUsage(varRef, info.GetInputType()))
-                );
-            });
+            var listener = new MatchingNodeVisitor<VariableReference>(
+                (varRef, __) => usages.Add(new VariableUsage(varRef, info.GetInputType()))
+            );
 
             var visitor = new BasicVisitor(info, listener);
-            visitor.Visit(node);
+            visitor.Visit(node, null);
```
Comment (sungam3r) [inline src/GraphQL/Validation/ValidationContext.cs:80]: Why null?

**Comment by Shane32** [comment]
Suggest closing in favor of #2139

**Comment by sungam3r** [comment]
As you wish.

File: src/GraphQL/Validation/ValidationContext.cs (hunk 2 above)
Comment (Shane32) [reply src/GraphQL/Validation/ValidationContext.cs:80]: The context was not used in the lambda (the code is up just a few lines), nor was it available in the method.

**Comment by Shane32** [comment]
> Well I see that the idea of this PR the same as #2134. I see that in some places you don't use statics - where there is state. I went further and transferred the state to `ValidationContext`.

Yeah I think most of our code is exactly the same.

**Comment by sungam3r** [comment]
Closing?

**Comment by Shane32** [comment]
Closed in favor of #2139

**Pull request closed without merging**